rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/dap/RascalDebugAdapter.java

```diff
@@ -0,0 +1,420 @@
+        source.setPath(path);
+        source.setSourceReference(-1);
+        return source;
+    }
+
+    @Override
+    public CompletableFuture<ScopesResponse> scopes(ScopesArguments args) {
+        return CompletableFuture.supplyAsync(() -> {
+            int frameId = args.getFrameId();
+
+            List<Scope> scopes = new ArrayList<>();
+
+            IRascalFrame frame = suspendedState.getCurrentStackFrames()[frameId];
+            ScopesResponse response = new ScopesResponse();
+            Scope scopeLocals = new Scope();
+            scopeLocals.setName("Locals");
+            scopeLocals.setNamedVariables(frame.getFrameVariables().size());
+            scopeLocals.setPresentationHint("locals");
+            scopeLocals.setExpensive(frame.getFrameVariables().size()>expensiveScopeMinSize);
+            scopeLocals.setVariablesReference(suspendedState.addScope(frame));
+            scopes.add(scopeLocals);
+
+            for(String importName : frame.getImports()){
+                IRascalFrame module = evaluator.getModule(importName);
+
+                if(module != null && module.getFrameVariables().size() > 0){
+                    Scope scopeModule = new Scope();
+                    scopeModule.setName("Module " + importName);
+                    scopeModule.setNamedVariables(module.getFrameVariables().size());
+                    scopeModule.setPresentationHint("module");
+                    scopeModule.setExpensive(module.getFrameVariables().size()>expensiveScopeMinSize);
+                    scopeModule.setVariablesReference(suspendedState.addScope(module));
+                    scopes.add(scopeModule);
+                }
+            }
+
+            response.setScopes(scopes.toArray(new Scope[scopes.size()]));
+            return response;
+        });
+    }
+
+    @Override
+    public CompletableFuture<VariablesResponse> variables(VariablesArguments args) {
+        return CompletableFuture.supplyAsync(() -> {
+            int reference = args.getVariablesReference();
+            int minIndex = args.getStart() != null ? args.getStart() : 0;
+            int maxCount = args.getCount() != null ? args.getCount() : -1;
+
+            VariablesResponse response = new VariablesResponse();
+            List<RascalVariable> variables = suspendedState.getVariables(reference, minIndex, maxCount);
+            Variable[] variablesResponse = new Variable[variables.size()];
+            int i = 0;
+            for(RascalVariable var : variables){
+                Variable variable = new Variable();
+                variable.setName(var.getName());
+                variable.setType(var.getType().toString());
+                variable.setValue(var.getDisplayValue());
+                variable.setVariablesReference(var.getReferenceID());
+                variable.setNamedVariables(var.getNamedVariables());
+                variable.setIndexedVariables(var.getIndexedVariables());
+                variablesResponse[i] = variable;
```

Comment:
As an alternative (that is all up to you in to pick this up or not) we can also use streaming api to build an array.

`response.setVariables(variables.toStream().map(v -> newVariable(v)).toArray(Variable[]::new));` 

would be the single line that goes through all varialbes in the list, and maps it onto a new value (`Variable`) and then at the end collects it into a array. 

and if you want, the body of the for loop can be inline in the map callback.